// knip/dist/typescript/resolve-module-names.js
import { existsSync } from 'node:fs';
import { isBuiltin } from 'node:module';
import { DEFAULT_EXTENSIONS, DTS_EXTENSIONS } from '../constants.js';
import { sanitizeSpecifier } from '../util/modules.js';
import { timerify } from '../util/Performance.js';
import { dirname, extname, isAbsolute, isInNodeModules, join } from '../util/path.js';
import { _createSyncModuleResolver, _resolveModuleSync } from '../util/resolve.js';
function pickStringTarget(value) {
    if (typeof value === 'string')
        return value;
    if (!value || typeof value !== 'object' || Array.isArray(value))
        return;
    const obj = value;
    for (const key of ['default', 'node', 'import', 'require']) {
        const v = pickStringTarget(obj[key]);
        if (v)
            return v;
    }
    for (const v of Object.values(obj)) {
        const s = pickStringTarget(v);
        if (s)
            return s;
    }
}
const moduleResolutionCaches = [];
export function clearModuleResolutionCaches() {
    for (const cache of moduleResolutionCaches)
        cache.clear();
}
function compilePathMappings(scopedPaths) {
    if (!scopedPaths)
        return undefined;
    const mappings = [];
    for (const { scope, paths } of scopedPaths) {
        for (const key in paths) {
            const starIndex = key.indexOf('*');
            if (starIndex >= 0) {
                mappings.push({ prefix: key.slice(0, starIndex), wildcard: true, values: paths[key], scope });
            }
            else {
                mappings.push({ prefix: key, wildcard: false, values: paths[key], scope });
            }
        }
    }
    if (mappings.length === 0)
        return undefined;
    mappings.sort((a, b) => b.scope.length - a.scope.length);
    return mappings;
}
function compileRootDirs(scopedRootDirs) {
    if (!scopedRootDirs)
        return undefined;
    const scoped = scopedRootDirs.filter(({ rootDirs }) => rootDirs.length > 1);
    if (scoped.length === 0)
        return undefined;
    scoped.sort((a, b) => b.scope.length - a.scope.length);
    return scoped;
}
export function createCustomModuleResolver(compilerOptions, customCompilerExtensions, toSourceFilePath, findWorkspaceManifestImports, tsConfigFile) {
    const customCompilerExtensionsSet = new Set(customCompilerExtensions);
    const hasCustomExts = customCompilerExtensionsSet.size > 0;
    const extensions = [...DEFAULT_EXTENSIONS, ...customCompilerExtensions, ...DTS_EXTENSIONS, '.json', '.jsonc'];
    const resolveSync = hasCustomExts || tsConfigFile ? _createSyncModuleResolver(extensions, tsConfigFile) : _resolveModuleSync;
    const pathMappings = compilePathMappings(compilerOptions.scopedPaths);
    const rootDirMappings = compileRootDirs(compilerOptions.scopedRootDirs);
    function toSourcePath(resolvedFileName) {
        if (!hasCustomExts || !customCompilerExtensionsSet.has(extname(resolvedFileName))) {
            return toSourceFilePath(resolvedFileName) || resolvedFileName;
        }
        return resolvedFileName;
    }
    function toResult(resolvedFileName) {
        const mapped = toSourcePath(resolvedFileName);
        return {
            resolvedFileName: mapped,
            isExternalLibraryImport: mapped === resolvedFileName && isInNodeModules(resolvedFileName),
        };
    }
    const cache = new Map();
    moduleResolutionCaches.push(cache);
    function resolveModuleName(name, containingFile) {
        const dir = dirname(containingFile);
        let byName = cache.get(dir);
        if (byName) {
            if (byName.has(name))
                return byName.get(name);
        }
        else {
            byName = new Map();
            cache.set(dir, byName);
        }
        const result = resolveModuleNameUncached(name, containingFile);
        byName.set(name, result);
        return result;
    }
    function resolveModuleNameUncached(name, containingFile) {
        const specifier = sanitizeSpecifier(name);
        if (isBuiltin(specifier))
            return undefined;
        const resolvedFileName = resolveSync(specifier, containingFile);
        if (resolvedFileName)
            return toResult(resolvedFileName);
        if (pathMappings) {
            const dir = dirname(containingFile);
            for (const { prefix, wildcard, values, scope } of pathMappings) {
                if (dir !== scope && !dir.startsWith(`${scope}/`))
                    continue;
                if (wildcard ? specifier.startsWith(prefix) : specifier === prefix) {
                    const captured = wildcard ? specifier.slice(prefix.length) : '';
                    for (const value of values) {
                        const starIdx = value.indexOf('*');
                        const mapped = starIdx >= 0 ? value.slice(0, starIdx) + captured + value.slice(starIdx + 1) : value;
                        const resolved = resolveSync(mapped, containingFile);
                        if (resolved)
                            return toResult(resolved);
                    }
                }
            }
        }
        if (rootDirMappings && !isAbsolute(specifier)) {
            const dir = dirname(containingFile);
            for (const { scope, rootDirs } of rootDirMappings) {
                if (dir !== scope && !dir.startsWith(`${scope}/`))
                    continue;
                for (const rootDir of rootDirs) {
                    if (dir !== rootDir && !dir.startsWith(`${rootDir}/`))
                        continue;
                    const relPath = dir === rootDir ? '' : dir.slice(rootDir.length + 1);
                    for (const targetRoot of rootDirs) {
                        if (targetRoot === rootDir)
                            continue;
                        const resolved = resolveSync(join(targetRoot, relPath, specifier), containingFile);
                        if (resolved)
                            return toResult(resolved);
                    }
                }
            }
        }
        if (specifier.startsWith('#') && findWorkspaceManifestImports) {
            const ws = findWorkspaceManifestImports(containingFile);
            if (ws) {
                const target = pickStringTarget(ws.imports[specifier]);
                if (target?.startsWith('.')) {
                    const sourcePath = toSourceFilePath(join(ws.dir, target));
                    if (sourcePath)
                        return toResult(sourcePath);
                }
            }
        }
        const candidate = isAbsolute(specifier) ? specifier : join(dirname(containingFile), specifier);
        if (existsSync(candidate)) {
            return { resolvedFileName: candidate, isExternalLibraryImport: false };
        }
    }
    return timerify(resolveModuleName);
}
