// @tanstack/eslint-plugin-query/build/modern/index.js
import {
  rules
} from "./chunk-VALYN6GR.js";

// src/index.ts
var recommendedRules = {
  "@tanstack/query/exhaustive-deps": "error",
  "@tanstack/query/no-rest-destructuring": "warn",
  "@tanstack/query/stable-query-client": "error",
  "@tanstack/query/no-unstable-deps": "error",
  "@tanstack/query/infinite-query-property-order": "error",
  "@tanstack/query/no-void-query-fn": "error",
  "@tanstack/query/mutation-property-order": "error"
};
var recommendedStrictRules = {
  ...recommendedRules,
  "@tanstack/query/prefer-query-options": "error"
};
var plugin = {
  meta: {
    name: "@tanstack/eslint-plugin-query"
  },
  configs: {
    recommended: {
      plugins: ["@tanstack/query"],
      rules: recommendedRules
    },
    recommendedStrict: {
      plugins: ["@tanstack/query"],
      rules: recommendedStrictRules
    },
    "flat/recommended": [
      {
        name: "tanstack/query/flat/recommended",
        plugins: {
          "@tanstack/query": {}
          // Assigned after plugin object created
        },
        rules: recommendedRules
      }
    ],
    "flat/recommended-strict": [
      {
        name: "tanstack/query/flat/recommended-strict",
        plugins: {
          "@tanstack/query": {}
          // Assigned after plugin object created
        },
        rules: recommendedStrictRules
      }
    ]
  },
  rules
};
plugin.configs["flat/recommended"][0].plugins["@tanstack/query"] = plugin;
plugin.configs["flat/recommended-strict"][0].plugins["@tanstack/query"] = plugin;
var index_default = plugin;
export {
  index_default as default,
  plugin
};
