// unbash/dist/chars.js
export const CH_TAB = 9;
export const CH_NL = 10;
export const CH_SPACE = 32;
export const CH_BANG = 33;
export const CH_DQUOTE = 34;
export const CH_HASH = 35;
export const CH_DOLLAR = 36;
export const CH_PERCENT = 37;
export const CH_AMP = 38;
export const CH_SQUOTE = 39;
export const CH_LPAREN = 40;
export const CH_RPAREN = 41;
export const CH_STAR = 42;
export const CH_PLUS = 43;
export const CH_COMMA = 44;
export const CH_DASH = 45;
export const CH_DOT = 46;
export const CH_SLASH = 47;
export const CH_0 = 48;
export const CH_9 = 57;
export const CH_COLON = 58;
export const CH_SEMI = 59;
export const CH_LT = 60;
export const CH_EQ = 61;
export const CH_GT = 62;
export const CH_QUESTION = 63;
export const CH_AT = 64;
export const CH_A = 65;
export const CH_Z = 90;
export const CH_LBRACKET = 91;
export const CH_BACKSLASH = 92;
export const CH_RBRACKET = 93;
export const CH_CARET = 94;
export const CH_UNDERSCORE = 95;
export const CH_BACKTICK = 96;
export const CH_a = 97;
export const CH_z = 122;
export const CH_LBRACE = 123;
export const CH_PIPE = 124;
export const CH_RBRACE = 125;
export const CH_TILDE = 126;
